vms/platformvm/validators/fee/fee.go

```diff
@@ -11,34 +11,44 @@ import (
 	safemath "github.com/ava-labs/avalanchego/utils/math"
 )
 
-// CalculateExcess updates the excess value based on the target and current
-// usage over the provided duration.
-func CalculateExcess(
-	target gas.Gas,
-	current gas.Gas,
-	excess gas.Gas,
-	duration uint64,
-) gas.Gas {
-	if current < target {
-		return excess.SubPerSecond(target-current, duration)
+// Config contains all the static parameters of the dynamic fee mechanism.
+type Config struct {
+	Target                   gas.Gas   `json:"target"`
+	MinPrice                 gas.Price `json:"minPrice"`
+	ExcessConversionConstant gas.Gas   `json:"excessConversionConstant"`
+}
+
+// State represents the current on-chain values used in the dynamic fee
+// mechanism.
+type State struct {
+	Current gas.Gas `json:"current"`
+	Excess  gas.Gas `json:"excess"`
+}
+
+// AdvanceTime adds (s.Current - target) * seconds to Excess.
+//
+// If Excess would underflow, it is set to 0.
+// If Excess would overflow, it is set to MaxUint64.
+func (s State) AdvanceTime(target gas.Gas, seconds uint64) State {
+	excess := s.Excess
+	if s.Current < target {
+		excess = excess.SubPerSecond(target-s.Current, seconds)
+	} else if s.Current > target {
+		excess = excess.AddPerSecond(s.Current-target, seconds)
+	}
+	return State{
```

Comment:
Is this because we were talking about functional languages? 🤣

Reply:
It's actually because we might be passing these through different block states... And we definitely don't want them to be mutable there